src/DocumentDB.ChangeFeedProcessor/PartitionControllerHealthnessEvaluator.cs

```diff
@@ -0,0 +1,60 @@
+﻿//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  Licensed under the MIT license.
+//----------------------------------------------------------------
+
+namespace Microsoft.Azure.Documents.ChangeFeedProcessor
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.PartitionManagement;
+
+    internal class PartitionControllerHealthnessEvaluator : IPartitionController
+    {
+        private static readonly long UninitializedTicks = DateTime.MaxValue.Ticks;
+        private readonly IPartitionController inner;
+        private readonly long maxAllowedUnhealthyDuration;
+        private readonly IUnhealthyHandlingStrategy unhealthyStrategy;
+        private long firstUnhealthyTick = UninitializedTicks;
+
+        public PartitionControllerHealthnessEvaluator(IPartitionController inner, TimeSpan leaseExpirationTimeout, IUnhealthyHandlingStrategy unhealthyStrategy)
+        {
+            this.inner = inner ?? throw new ArgumentNullException(nameof(inner));
+            this.unhealthyStrategy = unhealthyStrategy ?? throw new ArgumentNullException(nameof(unhealthyStrategy));
+            this.maxAllowedUnhealthyDuration = leaseExpirationTimeout.Ticks * 5;
+        }
+
+        public async Task AddOrUpdateLeaseAsync(ILease lease)
+        {
+            try
+            {
+                await this.inner.AddOrUpdateLeaseAsync(lease);
+                Interlocked.CompareExchange(ref this.firstUnhealthyTick, UninitializedTicks, this.firstUnhealthyTick);
+            }
+            catch (DocumentClientException)
+            {
+                throw;
+            }
+            catch (Exception)
```

Comment:
We could send the Exception to the unhealthyStrategy.HandleAsync for logging or doing something particular depending on the type of exception?

Reply:
yeap

+ lease
+ maybe also processor instance?

Reply:
done